### PR TITLE
Test: MockScoreScript can be cacheable

### PR DESCRIPTION
**Pull request by stu-elastic**

### CODE REVIEW

**Review by jdconrad — APPROVED** [review]
Script side of this looks good to me. Thanks for fixing. Please have @mayya-sharipova review the changes to FunctionScoreQueryBuilderTests.

File: server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java [hunk 1]
```diff
@@ -837,21 +837,22 @@ public void testCacheability() throws IOException {
...
-        for (ScoreFunctionBuilder<?> scoreFunction : List.of(scriptScoreFunction, randomScoreFunctionBuilder)) {
-            FilterFunctionBuilder[] functions = new FilterFunctionBuilder[] {
-                    new FilterFunctionBuilder(RandomQueryBuilder.createQuery(random()), scoreFunction) };
-            queryBuilder = new FunctionScoreQueryBuilder(functions);
+        queryBuilder = new FunctionScoreQueryBuilder(new FilterFunctionBuilder[] {
+            new FilterFunctionBuilder(RandomQueryBuilder.createQuery(random()), scriptScoreFunction) });
+        context = createShardContext();
+        rewriteQuery = rewriteQuery(queryBuilder, new QueryShardContext(context));
+        assertNotNull(rewriteQuery.toQuery(context));
+        assertTrue("script should query should be cacheable: " + queryBuilder.toString(), context.isCacheable());
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java:847]: "script should query should be cacheable: "  -> "function script query should be cacheable"

File: server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java [hunk 2]
```diff
@@ -837,21 +837,22 @@ public void testCacheability() throws IOException {
...
-            assertNotNull(rewriteQuery.toQuery(context));
-            assertFalse("query should not be cacheable: " + queryBuilder.toString(), context.isCacheable());
-        }
+        RandomScoreFunctionBuilder randomScoreFunctionBuilder = new RandomScoreFunctionBuilderWithFixedSeed();
+        queryBuilder = new FunctionScoreQueryBuilder(new FilterFunctionBuilder[] {
+            new FilterFunctionBuilder(RandomQueryBuilder.createQuery(random()), randomScoreFunctionBuilder) });
+        context = createShardContext();
+        rewriteQuery = rewriteQuery(queryBuilder, new QueryShardContext(context));
+        assertNotNull(rewriteQuery.toQuery(context));
+        assertFalse("random score should query should not be cacheable: " + queryBuilder.toString(), context.isCacheable());
```
Comment (mayya-sharipova) [inline server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java:855]: "random score should query should not be cacheable: "  -> "function random query should not be cacheable: "

Comment (mayya-sharipova) [reply server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java:855]: We also need to remove `ScriptScoreFunctionBuilder` as a condition for `false` in the `isCacheable` function (line 861-862), as for `ScriptScoreFunctionBuilder` case we want to return `true`.

**Review by mayya-sharipova — APPROVED** [review]
@stu-elastic Thanks for the change, nice for the tests to have the same logic as in the common case.
I have left a couple of small comments

File: server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java (hunk 1 above)
Comment (stu-elastic) [reply server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java]: Added.

File: server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java (hunk 2 above)
Comment (stu-elastic) [reply server/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilderTests.java]: Updated message and removed `ScriptScoreFunctionBuilder`.

**Comment by stu-elastic** [comment]
master: 0ed1eb5
7.x: 93a2e9b

**Pull request merged**